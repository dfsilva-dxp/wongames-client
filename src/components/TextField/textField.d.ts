import { InputHTMLAttributes, ReactNode } from "react";

export type TextFieldProps = {
  // eslint-disable-next-line no-unused-vars
  onInput?: (value: string) => void;
  label?: string;
  initialValue?: string;
  icon?: ReactNode;
  iconPosition?: "left" | "right";
  disabled?: boolean;
  error?: string;
} & InputHTMLAttributes<HTMLInputElement>;
