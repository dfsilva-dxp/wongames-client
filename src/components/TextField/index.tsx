import { ChangeEvent, useState } from "react";

import * as S from "./styles";

import { TextFieldProps } from "./textField";

const TextField = ({
  label,
  initialValue = "",
  icon,
  name,
  iconPosition = "left",
  disabled = false,
  onInput,
  error,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue);

  const onChange = (e: ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value;
    setValue(newValue);

    !!onInput && onInput(newValue);
  };

  return (
    <S.Wrapper disabled={disabled} error={!!error}>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      <S.InputWrapper>
        {!!icon && <S.Icon iconPosition={iconPosition}>{icon}</S.Icon>}
        <S.Input
          type="text"
          onChange={onChange}
          value={value}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </S.InputWrapper>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  );
};

export default TextField;
