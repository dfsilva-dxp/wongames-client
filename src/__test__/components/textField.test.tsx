import { act, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { EmailOutline } from "@styled-icons/evaicons-outline/EmailOutline";

import TextField from "components/TextField";

import { renderWithTheme } from "utils/tests/helpers";

describe("<TextField />", () => {
  it("should renders with label", () => {
    const { container } = renderWithTheme(
      <TextField label="Label" name="Field" />
    );

    expect(screen.getByLabelText("Label")).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should renders without label", () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByLabelText("Label")).not.toBeInTheDocument();
  });

  it("should renders with placeholder", () => {
    renderWithTheme(<TextField placeholder="hey you" />);

    expect(screen.getByPlaceholderText("hey you")).toBeInTheDocument();
  });

  it("should changes its value when typing", async () => {
    const onInput = jest.fn();

    renderWithTheme(<TextField onInput={onInput} label="Label" name="Label" />);

    const input = screen.getByRole("textbox");
    const text = "This is my new text";

    act(() => {
      userEvent.type(input, text);
    });

    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(onInput).toHaveBeenCalledTimes(text.length);
    });

    expect(onInput).toHaveBeenCalledWith(text);
  });

  it("should is accessible by tab", () => {
    renderWithTheme(<TextField label="Label" name="Label" />);

    const input = screen.getByLabelText("Label");

    expect(document.body).toHaveFocus();

    act(() => {
      userEvent.tab();
    });

    expect(input).toHaveFocus();
  });

  it("should render an icon", () => {
    renderWithTheme(<TextField icon={<EmailOutline data-testid="icon" />} />);

    expect(screen.getByTestId("icon")).toBeInTheDocument();
  });

  it("should render an iconon the right side", () => {
    renderWithTheme(
      <TextField
        icon={<EmailOutline data-testid="icon" />}
        iconPosition="right"
      />
    );

    expect(screen.getByTestId("icon").parentElement).toHaveStyle({
      order: 1
    });
  });

  it("should does not changes its value when disabled", async () => {
    const onInput = jest.fn();
    renderWithTheme(
      <TextField
        onInput={onInput}
        label="TextField"
        name="TextField"
        disabled
      />
    );

    const input = screen.getByRole("textbox");
    expect(input).toBeDisabled();

    const text = "This is my new text";

    act(() => {
      userEvent.type(input, text);
    });

    await waitFor(() => {
      expect(input).not.toHaveValue(text);
    });
    expect(onInput).not.toHaveBeenCalled();
  });

  it("should is not accessible by tab when disabled", () => {
    renderWithTheme(<TextField label="TextField" name="TextField" disabled />);

    const input = screen.getByLabelText("TextField");
    expect(document.body).toHaveFocus();

    act(() => {
      userEvent.tab();
    });

    expect(input).not.toHaveFocus();
  });

  it("should renders with error", () => {
    const { container } = renderWithTheme(
      <TextField
        icon={<EmailOutline data-testid="icon" />}
        label="TextField"
        name="TextField"
        error="Error message"
      />
    );

    expect(screen.getByText("Error message")).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
