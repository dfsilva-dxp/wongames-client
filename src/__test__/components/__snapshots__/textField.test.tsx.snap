// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should renders with error 1`] = `
.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 .c3 {
  border-color: #FF6347;
}

.c0 .c5,
.c0 .c1 {
  color: #FF6347;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 1.6rem;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c8 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c2 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 2.2rem;
  color: #8F8F8F;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c6 > svg {
  width: 100%;
}

.c9 {
  color: #FF6347;
  font-size: 1.2rem;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="TextField"
  >
    TextField
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="c5 c6"
    >
      <svg
        aria-hidden="true"
        class="c7"
        data-testid="icon"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          data-name="Layer 2"
        >
          <path
            d="M19 4H5a3 3 0 0 0-3 3v10a3 3 0 0 0 3 3h14a3 3 0 0 0 3-3V7a3 3 0 0 0-3-3zm-.67 2L12 10.75 5.67 6zM19 18H5a1 1 0 0 1-1-1V7.25l7.4 5.55a1 1 0 0 0 .6.2 1 1 0 0 0 .6-.2L20 7.25V17a1 1 0 0 1-1 1z"
            data-name="email"
          />
        </g>
      </svg>
    </div>
    <input
      class="c8"
      id="TextField"
      name="TextField"
      type="text"
      value=""
    />
  </div>
  <p
    class="c9"
  >
    Error message
  </p>
</div>
`;

exports[`<TextField /> should renders with label 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 1.6rem;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
}

.c1:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c2 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c0 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

<div
  class=""
>
  <label
    class="c0"
    for="Field"
  >
    Label
  </label>
  <div
    class="c1"
  >
    <input
      class="c2"
      id="Field"
      name="Field"
      type="text"
      value=""
    />
  </div>
</div>
`;
